KEP-2535: update the KEP with on-by-default strategy

stlaz opened the pull request:
- One-line PR description:
This PR modifies the original KEP according to recent discussions between sig-node and sig-auth.

- Issue link: https://github.com/kubernetes/enhancements/issues/2535

- Other comments:
  - Modify the strategy of the KEP from a feature-gated behavior goal to aim for an on-by-default feature.
  - Remove the periodical cache cleanup mentions, defer it for a possible follow-up KEP.
  - Stress that no unnecessary pulls should be performed so that registries' availability does not become blocking when it shouldn't be.
  - Add TODOs for unresolved talking points.
  
/cc mikebrow haircommander enj liggitt pacoxu sairameshv

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -114,12 +117,16 @@ EnsureImagesExist step the kubelet will require authentication of present images
 not require authentication.
 
+The new behavior is designed in a way so that it replaces current behavior - it is going
+to be an on-by-default feature.
```
nit: clarify that it will be on-by-default after it graduates, this isn't proposing direct-to-beta-and-enabled

```suggestion
to be an on-by-default feature once graduated.
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +136,25 @@ use un-encrypted...
-kubelet, of container images that required authentication and a list of the
-authentications that successfully pulled the image.
-For beta an API will be considered to manage the ensure metadata.
+A new API will be created to represent metadata about image pulls being authenticated. This API
```
This makes it sound like there will be REST API surface, which isn't what we're proposing.

I think it's helpful to start by describing what the kubelet will do:

> The kubelet will track container images that required authentication and a list of the
authentications that successfully pulled the image. This list will persist across reboots of host and restart of
kubelet.

And then go into *how* it will do it

> This will be done using a files persisted within the kubelet directory on the node.
> The files will be structured and versioned using standard config file API versioning techniques.

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +136,25 @@ use un-encrypted...
-    - A boolean that toggles this behavior. If `false`, the kubelet will fallback to the
+A new kubelet configuration option will be added, as well as a feature
+gate to gate its use:
+- `noPullImageSecretRecheck`
```
let's noodle a little bit on whether we can come up with a config knob / knobs that makes the spectrum of enforcement coherent:

How would each of these use cases be expressed in the kubelet config?
```
1. Do not allow access to _any_ pre-loaded images without verification
2. Only allow access to a _specific_ list of pre-loaded images without verification
3. Allow access to _all_ pre-loaded and pulled-without-pull-secrets images without verification (default behavior when this feature is enabled)
4. Allow access to all images _without verification_ (default behavior when this feature is not enabled, not recommended)
```

(again, I'm not 100% sure we want to make number 4 possible)

strawman possibilities (I didn't spend a ton of time on this, just wanted to start you thinking):

For 1:
```yaml
preexistingImagePolicy: VerifyAccess
```

For 2:
```yaml
preexistingImagePolicy: VerifyAccessUnlessAllowlisted
preexistingImageAllowlist:
- myregistry.example.com/myimage:mytag
- myregistry.example.com/myotherimage:*
- mypublicregistry.example.com/*
```

For 3 (default policy):
```yaml
preexistingImagePolicy: AllowAccessWithoutVerification
```

stlaz replied:
Given the scenarios we've got, that's actually a good option list. I think I'd probably only name it differently, maybe structure it a bit more for simpler config validation:
```yaml
disableImageSecretPullRecheck: true | false # in case we actually want to be able to disable this
preexistingImagesPullPolicy:
    type: AlwaysVerify | IgnoreListed | NeverVerify
    ignoreList: # specific to image tag / just an image + all tags / whole registry
    - myregistry.example.com/myimage:mytag
    - myregistry.example.com/myotherimage:*
    - mypublicregistry.example.com/*
```
```go
type KubeletConfiguration struct {
...
   DisableImageSecretPullRecheck bool // defaults to false
   PreexistingImagesPullPolicy *PreexistingImagesSecretPullPolicy
...
}

type PreexistingImageAccessPolicy string

const (
    NeverVerifyPreexistingImages PreexistingImageAccessPolicy = "NeverVerify" // the default
    IgnoreListedPreexistingImages = "IgnoreListed"
    AlwaysVerifyPreexistingImages = "AlwaysVerify"
)

type PreexistingImagesSecretPullPolicy struct {
    Type PreexistingImageAccessPolicy
    IgnoreList []string
}
```

stlaz replied:
This was further noodled about during a meeting with sig-node yesterday, I'll work in the sketch we came up with.

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -72,17 +74,18 @@ Thus even if the image is present, it may still be reauthenticated.
+**TODO**: describe the kubelet configuration option for allow-listing images
+
+The goal is to enable this new behavior by default with an option to revert to the old one by using a kubelet configuration
+field `noPullImageSecretRecheck`.
```
I restructured the section a bit but mostly tried to stick to your phrasing.

bridgetkromhout commented:
/assign stlaz

pacoxu commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +140,48 @@ use un-encrypted...
+- `imagePullCredentialsVerificationPolicy`
+    - An enum of 4 possible values:
+      - `NeverVerify`
+        - anyone on a node can use any image present on the node
```
Should we mention that this is the behavior before?

stlaz replied:
added a mention of that 👍

haircommander commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +143,49 @@ use un-encrypted...
 
 This feature will not change the behavior of pod with `ImagePullPolicy` `Always`.
 
+Enforcing periodical repull might be important in order to, for example, be able
```
nit: periodic*

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -114,12 +124,16 @@ EnsureImagesExist step the kubelet will require authentication of present images
-pulled through the kubelet with no `ImagePullSecrets`/authentication required will
+Images already present at boot or loaded externally to the kubelet will not require
+authentication unless configured otherwise.
+Images successfully pulled through the kubelet with no `ImagePullSecrets`/authentication required will
```
```suggestion
Images successfully pulled through the kubelet with no `ImagePullSecrets` authentication required will
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +143,49 @@ use un-encrypted...
+      - `NeverVerify`
+        - **original behavior before this feature**
+        - anyone on a node can use any image present on the node
+      - `NeverVerifyNodePreloadedImages`
```
no strong opinion, but I'm not sure what the `Node` qualifier is adding... thoughts on shortening this to:

```suggestion
      - `NeverVerifyPreloadedImages`
```

stlaz replied:
Looks like I even used the shorter version later in the text. I'll use it everywhere :+1:

haircommander commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+well-defined structure that follows standard config file API versioning techniques.
+This is done to keep track of the authenticated images between kubelet restarts.
+
+The max size of the cache scales with the number of unique cache entries * the number of unique images that have not been garbage collected.
```
I think the GC piece could be a bit more explicit:
```
Cache entries for an image will be deleted when the image is garbage collected. Thus, the max size of the cache scales with the number of unique cache entries * the number of unique images

```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +143,49 @@ use un-encrypted...
+    - registry.host/org/image-name
+    - registry.host/public-org/*
+    - registry.host/*
+  - notice registry.host/org/image-name:tag would be invalid as image name is considered the most granular possible configuration
```
```suggestion
  - Note: `registry.host/org/image-name:tag` or `registry.host/org/image-name@sha256:...` would be invalid as image name is considered the most granular possible configuration
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +143,49 @@ use un-encrypted...
+  - a string list to be used with the `NeverVerifyAllowlistedImages` image pull credentials verification policy
+  - images specified by this list will not require credential reverification if they are requested
+    by a pod with `IfNotPresent` or `Never` image pull policies
+  - the list accepts wildcard suffix in a string where `*` is the wildcard character
```
I also realized some registries have leading regionalized subdomains they might want to wildcard (`foo.example.com/...`, etc). I'm on the fence about including that or not in the initial revision, but I wouldn't object to simple prefix/suffix glob. Might be worth checking main registries to see if that is widespread

stlaz replied:
It looks like regionalized subdomains are quite common:
- https://cloud.ibm.com/docs/Registry?topic=Registry-registry_overview
- https://docs.oracle.com/en-us/iaas/Content/Registry/Concepts/registryprerequisites.htm
- https://cloud.google.com/artifact-registry/docs/transition/gcr-repositories#gcr-domain-support
- https://learn.microsoft.com/en-us/azure/container-registry/container-registry-geo-replication#example-use-case

I very briefly read the above documents. It seems to me that the use of regional registry is usually deliberate, and so maybe it should be in our config, too.

There was also a mention of infixed regional domains in the IBM document - see the "Private domain name" column in https://cloud.ibm.com/docs/Registry?topic=Registry-registry_overview#registry_regions_local

I'd suggest starting with suffixes for the first alpha, we can always modify the behavior to be looser.

haircommander commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+type ImagePullIntent struct {
+    metav1.TypeMeta
+
+    ImageSpec string `json:"imageSpec"` // the filename is a hash of this value
```
can you explain why the filename is different here ( I assume because for the first time an image is pulled we don't have the ID yet)

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+##### Failure modes:
+
+We should always fail safe. If there's an error reading from the disk, the record
+MUST be assumed as non-existent. This might cause an addition request to pull an image.
```
for clarity, does "non-existent" mean "treated as a preloaded image" or "treated as a kubelet-pulled image with unknown credential info"?

stlaz replied:
It should be the latter. I'll clarify that in the text.

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
````diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+thus the sanitization consists of replacing the `<hash-alg>:` with `<hash-alg>~`.
+
+```go
+type ImagePullIntent struct {
````
is it possible for multiple parallel pull attempts of the same image spec with different credentials to happen in the kubelet? if so, how are those represented in a single file?

liggitt replied:
Talking through this, if intent-to-pull *only* records the image, and we handle a crash mid-pull by treating everything with an intent-to-pull record as needing reverification, we don't need *any* credential info in the intent file, and multiple parallel pulls of the same image can just ensure the intent file exists.

On kubelet start, we'd still want to look at dangling intent-to-pull records, and check if images for those exist on the node.
1. For dangling intent-to-pull images which *do* exist, create digest entries for those images indicating reverification was needed (this implies the ImagePulledRecord must be able to express "needs reverification", similar to what we'll do when we cannot read/decode an ImagePulledRecord), then clean up intent-to-pull
2. For dangling intent-to-pull entries which *don't* have an image, we could leave the intent entries in place (either permanently or for "a while" ...), which will indicate future attempts to use those exact same image specs should be reverified

pacoxu replied:
If `NeverVerifyPreloadedImages`, dangling intent-to-pull entries which do exist, will become a preloaded image. How to distinguish between preloaded and dangling if we don't have the annotation?
- In the case of https://github.com/kubernetes/enhancements/pull/3547/files#diff-77a289ef32ead94beffe8193738841bf801db831a6383ed7de7342ccb0e1241c, sometimes user may trigger a manual image pull using crictl to pull manually.

If NeverVerifyAllowlistedImages(if not in the list), we can just do a reverification.

liggitt replied:
> dangling intent-to-pull entries which do exist, will become a preloaded image

that should not happen... a dangling intent-to-pull entry means the kubelet tried to pull and the image should be treated as a pulled-by-kubelet image with unknown credentials, which should require reverification on first use

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+    NodeWideCredentials bool `json:"nodeWideCredentials"`
+
+    // Anonymous is a flag denoting that the image was pulled without authentication.
+    // It is mutually exclusive to all the other fields of this structure.
```
```suggestion
    // If true, it is mutually exclusive to all the other fields of this structure.
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+    // Anonymous is a flag denoting that the image was pulled without authentication.
+    // It is mutually exclusive to all the other fields of this structure.
+    // +optional
+    Anonymous bool `json:"anonymous"`
```
```suggestion
    Anonymous bool `json:"omitempty,anonymous"`
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+    KubernetesSecretCoordinates map[KubernetesSecretCoordinates]ImagePullSecret `json:"kubernetesSecretCoordinates"`
+
+    // NodeWideCredentials is a flag denoting that the image was pulled with node-wide
+    // credentials
```
if true, is this mutually exclusive with the credentialHashes / kubernetesSecretCoordinates fields?

stlaz replied:
I think so - I'll add that in the comment.

If my understanding is correct, this field would be turned to `true` if the credentials were pulled from the node's [DockerKeyring](https://github.com/kubernetes/kubernetes/blob/19e8e59d06cbdd067a9b5c141b674b63a09d0e5c/pkg/credentialprovider/plugins.go#L50).

[KEP-4412: Projected Service Account Tokens for Kubelet Image Credential Providers](https://github.com/kubernetes/enhancements/pull/4846) suggests the DockerKeyring might change in a way that the credentials would no longer be node-wide for some providers, and so that's something to keep in mind here. I assume a new field would be introduced for those cases.

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+    // NodeWideCredentials is a flag denoting that the image was pulled with node-wide
+    // credentials
+    // +optional
+    NodeWideCredentials bool `json:"nodeWideCredentials"`
```
```suggestion
    NodeWideCredentials bool `json:"omitempty,nodeWideCredentials"`
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+//   an image. This assumes uninterrupted pulls if the same set of credentials is
+//   spread across the cluster.
+type ImagePullCredentials struct {
+    // CredentialHashes is an index of hashes of all the credentials that an image was pulled with
```
specify how the hash is derived? (algorithm, the content being hashed?)

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+    metav1.TypeMeta
+
+    ImageID string `json:"imageId"` // the filename is a filesystem sanitized image of this value
+    Credentials ImagePullCredentials `json:"credentials"`
```
could probably unnest ImagePullCredentials fields into the top-level ImagePulledRecord, I don't think we're using ImagePullCredentials elsewhere

then, an ImagePulledRecord that only contains `apiVersion, kind, imageID` would mean "this was pulled by kubelet and has no other info about what credentials are valid"

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+type ImagePulledRecord struct {
+    metav1.TypeMeta
+
+    ImageID string `json:"imageId"` // the filename is a filesystem sanitized image of this value
```
is there a clearer name to use for this?

Does this correspond to the image reference returned from CRI? If so, maybe call it image reference and document that connection?

```
	// PullImage pulls an image from the network to local storage using the supplied
	// secrets if necessary. It returns a reference (digest or ID) to the pulled image.
	PullImage(ctx context.Context, image ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, error)
	// GetImageRef gets the reference (digest or ID) of the image which has already been in
	// the local storage. It returns ("", nil) if the image isn't in the local storage.
	GetImageRef(ctx context.Context, image ImageSpec) (string, error)
```

Is CRI *required* to return the digest from PullImage / GetImageRef, or does it just *happen* to? I'm wondering how file-safe we can assume the image id is in terms of sanitizing it for a filename.

stlaz replied:
Image reference would likely be a better name for this field indeed.

I believe the CRI returns its own identification of the pulled image from `PullImage()`, which can either be directly the image repository digest, or the CRI internal ID of the image. I am no expert on container runtimes though so I'd rather have @haircommander or @mikebrow confirm my assumption.

haircommander replied:
yeah it's left this way historically, but ends up being an ImageRef. What further complicates the situation is containerd and cri-o do different things with the result of PullImage (containerd returns a digest, cri-o returns a repo+digest). Thus coming from cri-o I don't think we can assume it's file safe

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+1. kubelet prepares for a pull by writing an intent record
+2. kubelet issues an image pull request to the CRI
+3. kubelet restarts
+4. kubelet does not see the image so it removes the record from 1.
```
I was experimenting a little, but while I'm seeing the annotations being sent by the kubelet alongside the pull request, it doesn't seem that these would later be persisted - `crictl inspecti <imageref>` shows the imagespec, but the annotations are missing.

Perhaps @mikebrow might know if there's a way to do this, and if it would reliably work across different container runtimes.

liggitt replied:
Ok, worth following up but doesn't seem like we can rely on the annotations for now. The [other discussion](https://github.com/kubernetes/enhancements/pull/4789#discussion_r1754951743) about how to handle dangling intent-to-pull records simplifies and resolves the race between a background pull and a kubelet crash.

liggitt replied:
> sometimes user may trigger a manual image pull using crictl to pull manually or to see the progress.

out of band pulls the kubelet never attempted to pull would be treated like preloaded images, and are out of scope for this enhancement

mikebrow replied:
> I was experimenting a little, but while I'm seeing the annotations being sent by the kubelet alongside the pull request, it doesn't seem that these would later be persisted - `crictl inspecti <imageref>` shows the imagespec, but the annotations are missing.
> 
> Perhaps @mikebrow might know if there's a way to do this, and if it would reliably work across different container runtimes.

They should be stored in meta, if not it is an error.

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+type ImagePullCredentials struct {
+    // CredentialHashes is an index of hashes of all the credentials that an image was pulled with
+    // +optional
+    CredentialHashes map[string]ImagePullCredentials `json:"credentialHashes"`
```
Moved the hashes inside of a struct along with kube secret coordinates. I actually moved to a single value in the process. Do you think we should keep more per secret?

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+for credential reverification (and therefore ignored for caching).
+
+The rules that define what should be ignored are set by the `imagePullCredentialsVerificationPolicy`:
+- `NeverVerify` disables the cache completely (this was the previous behavior)
```
That's a very good point, I'll mention that the pull records should still be kept.

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+a repull if there's an attempt to fetch the same image with different, currently unknown credentials. In case the
+credentials that come with an image pull request were successfully used before, a repull at the registry must not be issued.
+
+Note that using the tag `:latest` is equivalent to using the image pull policy `Always.`
```
This was kept here from the original KEP but after looking into the code, it seems that the `:latest` tag really only has a meaning in the defaulting logic. I'll remove the sentence.

stlaz commented:
I believe the comments should now be all addressed. I added a comment in the API draft and kubelet housekeeping about what a "pulled" image record should look like to denote the need for creds reverification.

I also drafted an interface we could use for the image pull manager.

haircommander commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -382,58 +438,58 @@ container runtime. This KEP expects the image IDs provided to be in the form of
+    // If true, it is mutually exclusive with KubernetesSecrets and NodeWideCredentials.
     // +optional
-    Anonymous bool `json:"anonymous"`
+    Anonymous bool `json:"anonymous,omitempty"`
```
do we need to differentiate between these?

stlaz replied:
this is more of a cosmetic change, the deserialization stays the same

haircommander commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -288,6 +288,46 @@ There are two different stages when the kubelet updates its cache:
+    // the image appears on the node after the kubelet restarts.
+    RecordImagePulled(imageRef string, credentials *ImagePullCredentials) error
+    // RecordImagePullFailed should be called if an image failed to pull.
+    RecordImagePullFailed(image string) error
```
Maybe mention the action we take in this situation in the commends (which I presume is remove the intent record?)

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+Upon kubelet start, the `pulling` part of the cache should be checked for dangling items.
+These would be the records of images in the *pulling* state that now exist on the node
+as listed by the CRI. If such a record exists, a `pulled` record is created in
+its place with all authentication-related fields set to their default values,
```
```suggestion
its place with all authentication-related fields unset,
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+    // Returns true if policy demands reverification and no record of the pull
+    // was found in the cache.
+    ShouldReverifyCredentials(image string, credentials *ImagePullCredentials) bool
+}
```
need a method on this interface for the GC loop to remove a credential record for an image id, I think

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+    // set of credentials.
+    // Returns true if policy demands reverification and no record of the pull
+    // was found in the cache.
+    ShouldReverifyCredentials(image string, credentials *ImagePullCredentials) bool
```
maybe we want a different arg to ShouldReverify so we can make a single call instead of in a loop (and *shouldreverify* doesn't really need to explicitly opt into checking anonymous / node-wide)

liggitt replied:
doc the `image` arg corresponds to the pod spec

if we only call this when we've already resolved `image` → and existing `imageRef` on the node, should we pass that info in to this call so the implementation doesn't have to resolve it again?

pacoxu replied:
In the past implementation, we make a *credentialprovider.DockerKeyring from the pullSecrets.

https://github.com/kubernetes/kubernetes/blob/21f54c3ca1ac87611e101f46d8b05682f51efbee/pkg/kubelet/kuberuntime/kuberuntime_image.go#L42-L82

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+// One of the attributes must always be non-nil, and the others must be nil.
+type ImagePullCredentials struct {
+    KubernetesSecret *imagemanager.ImagePullSecret
+    NodeWideCredentials *bool
+    Anonymous *bool
```
can collapse these to a single field named something like `AllowAllPods *bool`

stlaz replied:
I started using `NodePodsAccessible` but I don't insist on it if `AllowAllPods` or any other feels more appropriate to the native speaker's ear (eye?).

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+    // ImageRef is a reference to the image represented by this file as received
+    // from the CRI.
+    // The filename is a filesystem-sanitized form of this value.
+    ImageRef string `json:"imageRef"`
```
maybe match the name the CRI uses to refer to this string (imageID, imageRef, etc)

stlaz replied:
looks like `imageRef` is used there, too

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+    // +optional
+    KubernetesSecrets []ImagePullSecret `json:"kubernetesSecretCoordinates"`
+
+    // NodeWideCredentials is a flag denoting that the image was pulled with node-wide
+    // credentials.
+    // If true, it is mutually exclusive with Anonymous and KubernetesSecrets.
+    // +optional
+    NodeWideCredentials bool `json:"nodeWideCredentials,omitempty"`
+
+    // Anonymous is a flag denoting that the image was pulled without authentication.
+    // If true, it is mutually exclusive with KubernetesSecrets and NodeWideCredentials.
+    // +optional
+    Anonymous bool `json:"anonymous,omitempty"`
```
can collapse these fields into a single `AllowAllPods bool` as well

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+    // The filename is a filesystem-sanitized form of this value.
+    ImageRef string `json:"imageRef"`
+
+    // KuberneteSecretCoordinates is an index of coordinates of all the kubernetes
```
nest / partition the credential info under a map keyed by the `registry/repo/image` that was pulled (can drop the tag and the digest from the image spec)

```
// map of imagespec (without tag/digest) to credentials valid for that imagespec
CredentialInfo map[string]CredentialInfo
```

This ensures that an anonymous pod can't reference a private registry or repo that *happens* to have a tagged public image in it, and discover that exists and what content it identifies.

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+type ImagePullManager interface {
+    // RecordPullIntent records an intent to pull an image.
+    // It returns a number of current pulls requesting the same image.
+    RecordPullIntent(image string) (uint23, error)
```
also make it very explicit what the image arg is (is it *exactly* what is in the pod spec?)

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+    // the image appears on the node after the kubelet restarts.
+    RecordImagePulled(imageRef string, credentials *ImagePullCredentials) error
+    // RecordImagePullFailed should be called if an image failed to pull.
+    RecordImagePullFailed(image string) error
```
same comment here about the arg being exactly what is in the pod spec

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+    // a reverification should be required upon next pull request. This can be
+    // useful in cases where the kubelet crashed in the middle of the pull but
+    // the image appears on the node after the kubelet restarts.
+    RecordImagePulled(imageRef string, credentials *ImagePullCredentials) error
```
I think we *have* to pass in the same `image string` arg we pass to `RecordPullIntent` so refcounting can get cleaned up properly, *and* so the pulled record can partition these credentials by the image that was being pulled (dropping any tag and digest from the image spec).

We *also* need the canonical `imageRef` string returned from CRI.

```go
// image corresponds to the pod spec and the image passed to RecordPullIntent.
// imageRef is the canonical pulled/resolved id returned from the CRI.
RecordImagePulled(image string, imageRef string, credentials *ImagePullCredentials) error
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+    // ShouldReverifyCredentials evaluates the policy for image specified in
+    // "image" and if the policy demands reverification, it checks the internal
+    // cache to see if there's a record of pulling the image with the presented
+    // set of credentials.
```
or pulled in a way all pods are allowed to use.

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+
+`ImagePullIntent` is the structure that is present in files in the `image_manager/pulling`
+directory. The filename of each such file is a SHA-256 hash of the image spec prepended
+by `sha256~`.
```
Let's go with the `-` then, to prevent unwanted expansions :+1:

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+
+`ImagePulledRecord` is the structure that appears in files in the `image_manager/pulled`
+directory. The filename of each such file is a sanitized image ID provided by the
+container runtime. This KEP expects the image IDs provided to be in the form of `<hash-alg>:<hash-val>`,
```
Moved the filenames to hashes, looks like that's the safest we can do :-|

stlaz commented:
I believe the comments from the last review round should now be addressed.

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+    // RecordImagePulled writes a record of an image being successfully pulled
+    // with ImagePullCredentials.
+    //
+    // If `credentials` is nil, it means the image was pulled successfully, but
+    // a reverification should be required upon next pull request. This can be
+    // useful in cases where the kubelet crashed in the middle of the pull but
+    // the image appears on the node after the kubelet restarts.
```
I don't think we'd expect to call this to deal with dangling intent records recovering from a crash... that would happen internally on construction.

I think every call to RecordImagePulled *must* provide a non-nil `credentials *imagemanager.ImagePullCredentials`, with *either* a single image pull secret OR NodePodsAccessible=true

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+    // the image appears on the node after the kubelet restarts.
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordImagePulled(image, imageRef string, credentials *imagemanager.ImagePullCredentials) error
```
we may not need the error return if this is inactionable for the caller and it is fine for this particular caller to use the successfully pulled image anyway

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+    // counter reaches zero, the pull intent record for the image is removed.
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordImagePullFailed(image string) error
```
we may not need the error return if this is inactionable for the caller and the caller already has a more important error to return (the image pull failed!)

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+    // was found in the cache.
+    //
+    // `image` is the content of the pod's container `image` field.
+    ShouldReverifyCredentials(image, imageRef string, credentials ... *imagemanager.ImagePullSecret) bool
```
there's internal verification that happens inside this method where the credentials are compared to the cache, but what this is *really* saying if it returns true is that a pull must be attempted

this might be a clearer name:

```suggestion
    MustAttemptImagePull(image, imageRef string, credentials []imagemanager.ImagePullSecret) bool
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+    //     ImagePulledRecord of an image it deleted from the node.
+    //     Using this method outside the GC context might cause security issues
+    //     if the image represented by the `imageRef` still exists on the node at
+    //     the time this method is called.
```
If we go with the `remove(imageRef)` approach, document this should be called *after* the image is actually removed successfully.

An alternative is to pass in all known image refs (which the kubelet / gc periodically list anyway) and let the implementation clean up pulled records that aren't in that list. That would self-heal the leaked record case eventually, but may introduce races between image pulls and image listing.

```
GC thread:
  T0: token = intenttoprune() // token could be a timestamp or a generation, etc
  T10: knownImages = list all images
  T20: maybe gc stuff
  T30: pruneunknownrecords(knownImages, token) // prunes to knownImages+things recorded since intenttoprune was called (based on token)

kubelet pull thread(s)
  T15: kubelet records intent to pull foo
  T25: kubelet successfully pulls foo and records success (in a way that pruneunknownrecords can fold in)
```

I think the paired IntentToPrune / PruneUnknownRecords approach where the full list of images known to GC is passed in works better.

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+    //     Using this method outside the GC context might cause security issues
+    //     if the image represented by the `imageRef` still exists on the node at
+    //     the time this method is called.
+    DestroyImagePulledRecord(imageRef string) error
```
I think the error is inactionable for the caller, consider dropping it?

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -458,24 +473,37 @@ type ImagePulledRecord struct {
+    //     "credentialMapping": {
+    //       "hello-world": { "nodePodsAccessible": true }
+    //     }
+    CredentialMapping map[string][]ImagePullCredentials `json:"credentialMapping,omitempty"`
```
should be just `CredentialMapping map[string]ImagePullCredentials`

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+    //     "credentialMapping": {
+    //       "hello-world": { "nodePodsAccessible": true }
+    //     }
+    CredentialMapping map[string][]ImagePullCredentials `json:"credentialMapping,omitempty"`
```
```suggestion
    CredentialMapping map[string]ImagePullCredentials `json:"credentialMapping,omitempty"`
```

pacoxu commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +143,49 @@ use un-encrypted...
+        - anyone on a node can use any image present on the node
+      - `NeverVerifyPreloadedImages`
+        - **default value**
+        - images that were pulled to the node by something else than the kubelet can be used without reverifying pull credentials
```
The statement here is different with below caching policy.

In line 270-271, the Preloaded images means: image pulled without a secret by kubelet + image pulled by other processes.

Here, we only mentioned  image pulled by other processes.

pacoxu replied:
My statement may be no accurate. 

- Does `image pulled without a secret by kubelet` equal to public images? public images do not need verify, which may be a common sense.

stlaz replied:
That's a good point, I think we changed this a little bit. I'll change the policy definitions around line ~270 to only mention things pulled outside the kubelet process.

pacoxu commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -309,6 +632,15 @@ At beta we will revisit if e2e buckets are warranted for e2e node, and after gat
+#### Beta
+
+- Kubelet cache is duplicated in memory and on disk. This is more resilient to I/O
+  operation errors when reading from disk.
```
IIUC,  for FG `KubeletEnsureSecretPulledImages`
- alpha: 
  - FG default false, imagePullCredentialsVerificationPolicy doesn't work. Behavior is same with `NeverVerify`.
  - When enabled, the imagePullCredentialsVerificationPolicy default is NeverVerifyPreloadedImages.
- beta: 
  - FG default true, imagePullCredentialsVerificationPolicy default to NeverVerifyPreloadedImages?
  - If you want to use the old behavior, change imagePullCredentialsVerificationPolicy=NeverVerify. Or you can just disable the FG. Either is OK.

There would be a behavior change when promoting the FG to beta.

pacoxu commented:
BTW, should this do a prod-readiness review again as there are many changes for the alpha design?
/cc @johnbelamaric
according to https://github.com/kubernetes/enhancements/blob/master/keps/prod-readiness/sig-node/2535.yaml

pacoxu commented:
/lgtm

pacoxu commented on keps/sig-node/2535-ensure-secret-pulled-images/kep.yaml:
```diff
@@ -16,11 +17,11 @@ approvers:
   - "@dchen1107"
   - "@derekwaynecarr"
```
Probably you should update the reviewer/approvers here.

pacoxu commented:
/assign @dchen1107 @derekwaynecarr

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -165,57 +201,365 @@ ensure all tenants have rights to the images that are already present on a host.
+    //
+    // This method is only expected to be called by the kubelet's image garbage collector.
+    // `until` is a timestamp created _before_ the `imageList` was requested from the CRI.
+    PruneUnknownRecords(imageList []string, until time.Time)
```
does the timestamp here mean we need to keep track of the time when a given imagePulledRecord was created? is the plan to do this in a field inside the record, or with a filesystem timestamp, or in memory?

stlaz replied:
I originally intended to have this be an implementation detail of the cache, which would imply the files' "modify" metadata information in the alpha case.
Should we explicitly track the `LastUpdatedTime metav1.Time` timestamp instead to avoid fs timestamp being unreliable?

liggitt replied:
I think it's probably better to make the update time explicit in the file

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +143,49 @@ use un-encrypted...
+      - `NeverVerifyAllowlistedImages`
+        - like `NeverVerifyPreloadedImages` but only node images from `preloadedImagesVerificationAllowlist` don't require reverification
+      - `AlwaysVerify`
+        - all images pulled with credentials require credential reverification
```
preloaded images / images without pull records would require verification as well ... I think this can simplify to:

```suggestion
        - all images require credential reverification
```

liggitt commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -129,27 +143,49 @@ use un-encrypted...
+  - a string list to be used with the `NeverVerifyAllowlistedImages` image pull credentials verification policy
+  - images specified by this list will not require credential reverification if they are requested
+    by a pod with `IfNotPresent` or `Never` image pull policies
+  - the list accepts wildcard suffix in a string where `*` is the wildcard character
```
do we intend to limit the wildcard to a full path segment, or are we intending to match `example.com/foo-*`?

if the former, should be something like:
```suggestion
  - the list accepts a `/*` wildcard suffix
```

liggitt reviewed (commented):
Just a couple questions, lgtm otherwise. Thanks for all the work on this

stlaz commented:
> Just a couple questions, lgtm otherwise. Thanks for all the work on this

Happy to be part of the process, these are some exciting changes :slightly_smiling_face: 

I addressed your comments. The timestamping is still an open question, see what you think. Either fs timestamps or a specific API field type sound good to me.

I also added a few small changes to the PRR questionnaire.

stlaz commented:
In the latest update I added `LastUpdatedTime metav1.Time` to the `ImagePulledRecord` structure, per https://github.com/kubernetes/enhancements/pull/4789#discussion_r1784498617

enj commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -332,7 +674,11 @@ TBD subsequent to alpha
+    record caching, which would make it easier to turn on the feature later.
   - Will enabling / disabling the feature require downtime of the control
     plane?
     - No, only a restart of the kubelet
```
Running pods will not be impacted though right?  A drain is not needed IIUC?

liggitt replied:
correct, this only kicks in during the ensureimage step

enj reviewed (commented):
Minor comments, LGTM.

stlaz commented on keps/sig-node/2535-ensure-secret-pulled-images/kep.yaml:
What kind of comment should it be? Won't the current changes to the file do?

liggitt replied:
nothing in this PR is touching keps/prod-readiness/sig-node/2535.yaml so nothing is *requiring* an approval... 

/assign @johnbelamaric 
to re-ack PRR for re-do for alpha, I guess

liggitt replied:
/unassign @johnbelamaric 
/assign @jpbetz 

@stlaz can you switch the alpha PRR review to @jpbetz to help load balance?

stlaz replied:
Done! Thank you Joe for volunteering on this one :slightly_smiling_face:

mrunalp commented:
Thanks for all the work that went into designing this :+1: 

/approve

mrunalp commented:
/hold 
For PRR

stlaz commented:
The latest change only switches the PRR approver and adds Jordan to reviewers.

pacoxu commented:
/lgtm
ping @jpbetz for PRR review

jpbetz commented on keps/sig-node/2535-ensure-secret-pulled-images/README.md:
```diff
@@ -350,8 +696,11 @@ Yes.
-Images pulled during the period the feature was disabled will not be present in the cache, and thus could incur redundant pulls/container creation failures.
-However, the cache may still be present, and thus it will retain information from when it was previously enabled.
+Images pulled during the period the feature was disabled will not be present in the cache. Depending on the policy configured,
+this could either incur redundant pulls/container creation failures OR would cause the feature to not work properly for
+images pulled during the time the feature was disabled.
```
When the policy is "AlwaysVerify" and registry is unavailable then the containers will not be able to create.  Clarify this more precisely?

"cause the feature to not work properly" - Clarify that the images pulled during this time will be treated.

stlaz replied:
Good suggestion, see if the latest change addresses your comment :+1:

jpbetz reviewed (commented):
Two PRR suggestions then LGTM

jpbetz commented:
/approve
For PRR

liggitt commented:
/hold cancel
/lgtm
